### PR TITLE
Support except and restrict in `push!()` and run()`

### PR DESCRIPTION
**Pull request by hhaensel**
Currently, it is not possible to address single clients by push! or run commands.

To remedy this situation I propose to
- modify existing keyword `except` to support not only UInt, but also Vetor{UInt} to exclude more than one client
- remove `WebSocket` as a possible value for except and rather supply the id as it is done in WebThreads
- add a keyword `restrict` to restrict messages to a client or list of clients

There will be a corresponding PR in Genie.

### CODE REVIEW

**Comment by hhaensel** [comment]
The current implementation for downloads is the following; in order to add the client info you add the keyword :addclient to the `@on` macro, e.g.

```julia
ui() = btn("Download", nocaps = true, icon = "download", @on(:click, :download_txt, :addclient))

@event :download_txt begin                                                                                                                                                                                                                                                 
    download_text(__model__, "Hello world"; client = event["_client"])                                                                                                                                                                                   end
```

**Comment by PGimenez** [comment]
If I understand correctly, this would allow us to trigger downloads from the backend. Right?

One user recently asked about this [on Discord](https://discord.com/channels/774897545717219328/1173673182079758440), and I came up with a MWE ([link](https://learn.genieframework.com/docs/examples/reactive-ui/trigger-download)) that does the following:
- Add a download link to the page
- Add a JS watcher to the link URL that clicks the link when the URL changes
- Change the link URL from a reactive handler to trigger the download without user click

I believe this PR would make all the added JS unnecessary

**Comment by essenciary** [comment]
"Trigger download from backend" -> that's simply a redirect isn't it? To the path/route that triggers the download...

**Comment by hhaensel** [comment]
In https://github.com/GenieFramework/StippleDemos/blob/master/Plugins/StippleDownloads.jl you find a demo app

**Comment by hhaensel** [comment]
@PGimenez @essenciary Do you think we should add client info per default, or should we keep it as optin by appending :addclient to the `@on` macro. Or should we just append `_addclient` to the eventkey
```julia
# 1. current implementation
ui() = btn("Download", nocaps = true, icon = "download", @on(:click, :download_txt, :addclient))

# 2. alternative implementation
ui() = btn("Download", nocaps = true, icon = "download", @on(:click, :download_txt_addclient))

# 3. always add client info (no change to current implementation)
ui() = btn("Download", nocaps = true, icon = "download", @on(:click, :download_txt))
```

**Comment by essenciary** [comment]
> > "Trigger download from backend" -> that's simply a redirect isn't it? To the path/route that triggers the download...
> 
> I don't register a route to download the content. I add an anchor element with a data-url that contains data from a data buffer. The data is sent beforehand with a push! command to the field '**download**' and then converted to a data buffer by `Uint8Array.from(this.__download__)`. Then the anchor element is clicked by a script to trigger the download. It doesn't need more than the above example to make personalised downloads available.

OK - but why? 

Literally all we had to do for that was have a route /__/download?file=abc -- and open it in a new tab.

Am I missing something?

**Comment by hhaensel** [comment]
One of my use cases is that users calculate an optimised mixture and store the result on their computer. The file never exists on the server.

**Comment by essenciary** [comment]
We don't need to register/deregister the route, there can always be a download utility there via Stipple.Downloads (eg on __init__ of the module). Then the file name can be encrypted/hashed. 

I did check the demo and hence my concern - it's not easy to follow and understand and it's quite complex. Hence my comment: is that the best/simplest way to achieve the objective? The code will need to be maintained for years to come.

**Comment by essenciary** [comment]
OK, I understand - for files that don't exist on the server, that use case makes sense. 
Also - is this for distinct plugin? If it's for a plugin (vs Stipple core) then I'm less concerned with complexity as it's optional addon.

File: src/Stipple.jl [hunk 1]
```diff
@@ -608,15 +615,16 @@ const max_retry_times = 10
...
     Base.push!(app::M, vals::Pair{Symbol,T}; channel::String,
-                except::Union{Genie.WebChannels.HTTP.WebSockets.WebSocket,Nothing,UInt}) where {T,M<:ReactiveModel}
+                except::Union{Nothing,UInt,Vector{UInt}}) where {T,M<:ReactiveModel}
 
 Pushes data payloads over to the frontend by broadcasting the `vals` through the `channel`.
 """
 function Base.push!(app::M, vals::Pair{Symbol,T};
-                    channel::String = Genie.config.webchannels_default_route,
-                    except::Union{Genie.WebChannels.HTTP.WebSockets.WebSocket,Nothing,UInt} = nothing)::Bool where {T,M<:ReactiveModel}
+                    channel::String = getchannel(model),
```
Comment (essenciary) [inline src/Stipple.jl:623]: Is `model` here correct? Where is it coming from?

**Comment by hhaensel** [comment]
It's our choice to include it in Stipple or to leave it as optional plugin.

The current PR is the basis for it to work, but there will be other uses as well, as you mentioned.

I did compared different implementations of js download. This seemed to me the most general one that can be adapted by users if they like. I don't think it will need lots of maintenance.

File: src/Stipple.jl (hunk 1 above)
Comment (hhaensel) [reply src/Stipple.jl:623]: no, it should be `app`

Comment (hhaensel) [reply src/Stipple.jl:623]: We had `Genie.config.webchannels_default_route` before and that generated errors, whenever I used it. I don't think that we still need the default route here, it's from the early days. Do you agree?
If yes, I'll fix that to be `app`

**Pull request merged**